File: frontend/utils/billing.ts
import axios from 'axios';
import { getSession } from './supabase';

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

// Log the API URL for debugging
console.log('API URL:', API_URL);

// Create axios instance with auth header
const api = axios.create({
  baseURL: API_URL,
  headers: {
    'Content-Type': 'application/json',
  },
});

// Add auth token to requests
api.interceptors.request.use(async (config) => {
  try {
    const session = await getSession();
    if (session?.access_token) {
      config.headers.Authorization = `Bearer ${session.access_token}`;
      console.log('Auth token added to request');
    } else {
      console.log('No auth token available');
    }
    
    // Log the full request URL for debugging
    console.log('Request URL:', `${config.baseURL}${config.url}`);
  } catch (error) {
    console.error('Error getting auth token:', error);
  }
  return config;
});

// Mock data for when the backend doesn't have billing endpoints
export const mockBillingData = {
  subscription: {
    plan: "Pro",
    status: "active",
    renewal_date: new Date().toISOString().split('T')[0],
    price: "$15.00",
    billing_cycle: "monthly",
    features: [
      "Unlimited images",
      "Up to 16x upscaling",
      "All upscaling modes",
      "API access",
      "Priority support",
    ]
  },
  payment_methods: [
    {
      id: "pm_1",
      type: "card",
      brand: "visa",
      last4: "4242",
      exp_month: 12,
      exp_year: 2024,
      is_default: true,
    }
  ],
  invoices: [
    {
      id: "in_1",
      date: new Date().toISOString().split('T')[0],
      amount: "$15.00",
      status: "paid",
      description: "Pro Plan - Monthly",
      download_url: "#",
    }
  ]
};

// Add cache system to reduce API calls
let billingInfoCache: any = null;
let billingInfoCacheTime = 0;
const CACHE_TTL = 5 * 60 * 1000; // 5 minutes

// Caching for billing history
let billingHistoryCache: any = null;
let billingHistoryCacheTime = 0;

// Convert storage_gb to storage_mb (100MB = 0.1GB, etc.)
// storage_gb: 0.1 → storage_mb: 100
// storage_gb: 5 → storage_mb: 5000
// storage_gb: 20 → storage_mb: 20000

// Billing API functions
export const getBillingInfo = async () => {
  try {
    // Check if we have a session before making the request
    const session = await getSession();
    if (!session) {
      console.error('No active session found');
      throw new Error('Authentication required');
    }
    
    // Check if we have cached data that's still valid
    const now = Date.now();
    if (billingInfoCache && (now - billingInfoCacheTime < CACHE_TTL)) {
      console.log('Using cached billing data');
      return billingInfoCache;
    }
    
    console.log('Making request to /billing endpoint');
    try {
      const response = await api.get('/billing');
      console.log('Response received:', response.status);
      
      // Cache the response
      billingInfoCache = response.data;
      billingInfoCacheTime = now;
      
      return response.data;
    } catch (error) {
      // If the endpoint doesn't exist (404), return mock data
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        console.log('Billing endpoint not found, using mock data');
        
        // Cache the mock data as well
        billingInfoCache = mockBillingData;
        billingInfoCacheTime = now;
        
        return mockBillingData;
      }
      throw error;
    }
  } catch (error) {
    console.error('Error fetching billing info:', error);
    
    // Add more detailed error logging
    if (axios.isAxiosError(error)) {
      console.error('Request URL:', error.config?.url);
      console.error('Request baseURL:', error.config?.baseURL);
      console.error('Response status:', error.response?.status);
      console.error('Response data:', error.response?.data);
    }
    
    throw error;
  }
};

export const updateSubscription = async (planId: string) => {
  try {
    try {
      const response = await api.post('/billing/subscription', { plan_id: planId });
      return response.data;
    } catch (error) {
      // If the endpoint doesn't exist (404), return mock success response
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        console.log('Subscription endpoint not found, using mock response');
        return {
          success: true,
          message: `Subscription updated to ${planId === 'pro' ? 'Pro' : planId === 'enterprise' ? 'Enterprise' : 'Free'} plan`,
          plan: {
            id: planId,
            name: planId === 'pro' ? 'Pro' : planId === 'enterprise' ? 'Enterprise' : 'Free',
            price: planId === 'pro' ? 15.00 : planId === 'enterprise' ? 30.00 : 0.00,
            interval: "month",
            features: planId === 'free' ? [
              "Up to 3 images per month",
              "2x and 4x upscaling",
              "Basic upscaling mode",
              "Standard support"
            ] : planId === 'pro' ? [
              "Unlimited images",
              "Up to 16x upscaling",
              "All upscaling modes",
              "API access",
              "Priority support"
            ] : [
              "Unlimited images",
              "Up to 16x upscaling",
              "All upscaling modes",
              "Unlimited API access",
              "Dedicated support",
              "Custom integration"
            ]
          }
        };
      }
      throw error;
    }
  } catch (error) {
    console.error('Error updating subscription:', error);
    throw error;
  }
};

export const cancelSubscription = async () => {
  try {
    try {
      const response = await api.post('/billing/subscription/cancel');
      return response.data;
    } catch (error) {
      // If the endpoint doesn't exist (404), return mock success response
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        console.log('Cancel subscription endpoint not found, using mock response');
        return {
          success: true,
          message: "Subscription cancelled successfully. You will have access until the end of your billing period."
        };
      }
      throw error;
    }
  } catch (error) {
    console.error('Error cancelling subscription:', error);
    throw error;
  }
};

export const addPaymentMethod = async (paymentDetails: any) => {
  try {
    try {
      const response = await api.post('/billing/payment-methods', paymentDetails);
      return response.data;
    } catch (error) {
      // If the endpoint doesn't exist (404), return mock success response
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        console.log('Payment methods endpoint not found, using mock response');
        return {
          success: true,
          message: "Payment method added successfully",
          payment_method: {
            id: "pm_new",
            type: "card",
            brand: "visa",
            last4: paymentDetails.card_number.slice(-4),
            exp_month: paymentDetails.exp_month,
            exp_year: paymentDetails.exp_year,
            is_default: false
          }
        };
      }
      throw error;
    }
  } catch (error) {
    console.error('Error adding payment method:', error);
    throw error;
  }
};

export const deletePaymentMethod = async (paymentMethodId: string) => {
  try {
    try {
      const response = await api.delete(`/billing/payment-methods/${paymentMethodId}`);
      return response.data;
    } catch (error) {
      // If the endpoint doesn't exist (404), return mock success response
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        console.log('Delete payment method endpoint not found, using mock response');
        return {
          success: true,
          message: "Payment method deleted successfully"
        };
      }
      throw error;
    }
  } catch (error) {
    console.error('Error deleting payment method:', error);
    throw error;
  }
};

export const setDefaultPaymentMethod = async (paymentMethodId: string) => {
  try {
    try {
      const response = await api.post(`/billing/payment-methods/${paymentMethodId}/default`);
      return response.data;
    } catch (error) {
      // If the endpoint doesn't exist (404), return mock success response
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        console.log('Set default payment method endpoint not found, using mock response');
        return {
          success: true,
          message: "Default payment method updated successfully"
        };
      }
      throw error;
    }
  } catch (error) {
    console.error('Error setting default payment method:', error);
    throw error;
  }
};

// Add a new function to get billing history
export const getBillingHistory = async () => {
  try {
    // Check if we have a session before making the request
    const session = await getSession();
    if (!session) {
      console.error('No active session found for billing history');
      throw new Error('Authentication required');
    }
    
    // Check if we have cached data that's still valid
    const now = Date.now();
    if (billingHistoryCache && (now - billingHistoryCacheTime < CACHE_TTL)) {
      console.log('Using cached billing history data');
      return billingHistoryCache;
    }
    
    // First try to get history from the billing endpoint
    console.log('Making request to /billing endpoint for history');
    try {
      const response = await api.get('/billing');
      console.log('Billing response received:', response.status);
      
      // Check if the billing response has invoices
      if (response.data && response.data.invoices && Array.isArray(response.data.invoices) && response.data.invoices.length > 0) {
        console.log('Found invoices in billing data');
        // Cache the response
        billingHistoryCache = response.data.invoices;
        billingHistoryCacheTime = now;
        return response.data.invoices;
      }
    } catch (error) {
      console.log('Could not get history from /billing endpoint, continuing to dedicated history endpoint');
      // Continue to next approach instead of throwing
    }
    
    // Next try the dedicated history endpoint
    console.log('Making request to /billing/history endpoint');
    try {
      const response = await api.get('/billing/history');
      console.log('Billing history response received:', response.status);
      
      // Return just the invoices array
      if (response.data && response.data.invoices && Array.isArray(response.data.invoices)) {
        return response.data.invoices;
      }
      
      // Backwards compatibility - if the endpoint returns a different format
      if (response.data && Array.isArray(response.data)) {
        return response.data;
      }
      
      // If response data exists but not in expected format, try to extract invoices
      if (response.data) {
        console.log('Trying to extract invoices from unexpected response format');
        for (const key in response.data) {
          if (Array.isArray(response.data[key]) && response.data[key].length > 0 && 
              response.data[key][0].hasOwnProperty('date') && response.data[key][0].hasOwnProperty('amount')) {
            console.log(`Found potential invoices array in key: ${key}`);
            return response.data[key];
          }
        }
      }
    } catch (error) {
      console.log('Billing history endpoint failed, trying Stripe API directly');
      // Continue to next approach instead of throwing
    }
    
    // Try fetching from Stripe customer portal invoices as last resort
    try {
      console.log('Making request to /stripe/invoices endpoint');
      const stripeResponse = await api.get('/stripe/invoices');
      console.log('Stripe invoices response received:', stripeResponse.status);
      
      if (stripeResponse.data && Array.isArray(stripeResponse.data)) {
        // Map Stripe invoice format to our format
        return stripeResponse.data.map((invoice: any) => ({
          id: invoice.id,
          date: new Date(invoice.created * 1000).toISOString().split('T')[0],
          amount: `$${(invoice.amount_paid / 100).toFixed(2)}`,
          status: invoice.status,
          description: invoice.description || `${invoice.billing_reason || 'Subscription'} payment`,
          downloadUrl: invoice.invoice_pdf || '#'
        }));
      }
      
      // If Stripe response data exists but not in expected format, look deeper
      if (stripeResponse.data) {
        console.log('Trying to extract invoices from unexpected Stripe response format');
        
        // Check if data.data is the array (common API pattern)
        if (stripeResponse.data.data && Array.isArray(stripeResponse.data.data)) {
          return stripeResponse.data.data.map((invoice: any) => ({
            id: invoice.id,
            date: new Date(invoice.created * 1000).toISOString().split('T')[0],
            amount: `$${(invoice.amount_paid / 100).toFixed(2)}`,
            status: invoice.status,
            description: invoice.description || `${invoice.billing_reason || 'Subscription'} payment`,
            downloadUrl: invoice.invoice_pdf || '#'
          }));
        }
        
        // Check other properties for arrays that might contain invoices
        for (const key in stripeResponse.data) {
          if (Array.isArray(stripeResponse.data[key]) && stripeResponse.data[key].length > 0) {
            if (stripeResponse.data[key][0].hasOwnProperty('created')) {
              console.log(`Found potential Stripe invoices array in key: ${key}`);
              return stripeResponse.data[key].map((invoice: any) => ({
                id: invoice.id,
                date: new Date(invoice.created * 1000).toISOString().split('T')[0],
                amount: `$${(invoice.amount_paid / 100).toFixed(2)}`,
                status: invoice.status,
                description: invoice.description || `${invoice.billing_reason || 'Subscription'} payment`,
                downloadUrl: invoice.invoice_pdf || '#'
              }));
            }
          }
        }
      }
    } catch (stripeError) {
      console.log('Stripe API not available, will return empty array');
      // Continue instead of throwing
    }
    
    // If all attempts fail but we have a session, create some simple mock data for display
    // This avoids showing "No billing history available" for paid users
    console.log('All billing history attempts failed, returning empty array');
    return [];
    
  } catch (error) {
    console.error('Error fetching billing history:', error);
    return [];
  }
};

export const getAvailablePlans = async () => {
  try {
    try {
      const response = await api.get('/billing/plans');
      return response.data;
    } catch (error) {
      // If the endpoint doesn't exist (404), return mock plans
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        console.log('Plans endpoint not found, using mock plans');
        return {
          plans: {
            free: {
              id: "free",
              name: "Free",
              price: 0.0,
              interval: "month",
              features: [
                "Up to 3 images per month",
                "2x and 4x upscaling",
                "Basic upscaling mode",
                "Standard support"
              ],
              limits: {
                images_per_month: 3,
                max_scale_factor: 4,
                api_calls_per_month: 0,
                storage_mb: 100
              }
            },
            pro: {
              id: "pro",
              name: "Pro",
              price: 15.0,
              interval: "month",
              features: [
                "Unlimited images",
                "Up to 16x upscaling",
                "All upscaling modes",
                "API access",
                "Priority support"
              ],
              limits: {
                images_per_month: 100,
                max_scale_factor: 16,
                api_calls_per_month: 500,
                storage_mb: 5000
              }
            },
            enterprise: {
              id: "enterprise",
              name: "Enterprise",
              price: 30.0,
              interval: "month",
              features: [
                "800 images per month",
                "Up to 16x upscaling",
                "All upscaling modes",
                "Unlimited API access",
                "Email support",
                "Custom integration"
              ],
              limits: {
                images_per_month: 800,
                max_scale_factor: 16,
                api_calls_per_month: 5000,
                storage_mb: 20000
              }
            }
          }
        };
      }
      throw error;
    }
  } catch (error) {
    console.error('Error fetching available plans:', error);
    throw error;
  }
}; 

File: backend/billing.py
from typing import Dict, List, Optional, Any
from pydantic import BaseModel
import stripe
from fastapi import HTTPException, status
import os
from datetime import datetime
import logging

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)

# Get Stripe API key from environment variables
stripe_api_key = os.getenv("STRIPE_SECRET_KEY")
if not stripe_api_key:
    logger.warning("STRIPE_SECRET_KEY environment variable is not set!")
else:
    logger.info(f"Stripe API key configured for billing.py (ending in: {stripe_api_key[-4:] if stripe_api_key else 'None'})")
    stripe.api_key = stripe_api_key

# Models for billing data
class SubscriptionPlan(BaseModel):
    id: str
    name: str
    price: float
    interval: str  # 'month' or 'year'
    features: List[str]
    limits: Dict[str, Any]

class PaymentMethod(BaseModel):
    id: str
    type: str
    brand: str
    last4: str
    exp_month: int
    exp_year: int
    is_default: bool

class Invoice(BaseModel):
    id: str
    date: str
    amount: float
    status: str
    description: str
    download_url: str

class UsageStats(BaseModel):
    images_processed: int
    images_limit: int
    api_calls: int
    api_calls_limit: int
    storage_used: str
    storage_limit: str

class BillingInfo(BaseModel):
    subscription: Optional[Dict[str, Any]] = None
    payment_methods: List[Dict[str, Any]] = []
    invoices: List[Dict[str, Any]] = []
    usage: Optional[Dict[str, Any]] = None

# Request model for checkout session creation
class CheckoutSessionRequest(BaseModel):
    plan_id: str
    price_id: str
    billing_cycle: str
    success_url: str
    cancel_url: str
    skip_auth: Optional[bool] = False

# Response model for checkout session
class CheckoutSessionResponse(BaseModel):
    url: str
    session_id: str

# Available subscription plans
SUBSCRIPTION_PLANS = {
    "free": SubscriptionPlan(
        id="free",
        name="Free",
        price=0.0,
        interval="month",
        features=[
            "5 images per month",
            "Basic upscaling (Only block mode)",
            "Maximum 2K output resolution",
            "Standard support"
        ],
        limits={
            "images_per_month": 5,
            "max_scale_factor": 2,
            "allowed_modes": ["block_mode"],
            "max_resolution": "2K",
            "api_calls_per_month": 0,
            "storage_mb": 100  # 0.1 GB = 100 MB
        }
    ),
    "pro": SubscriptionPlan(
        id="pro",
        name="Pro",
        price=15.0,
        interval="month",
        features=[
            "400 images per month",
            "Premium upscaling quality (Block mode, Face mode, and Waifu mode)",
            "Maximum 4K output resolution",
            "Fast processing speed",
            "Batch processing",
            "Email support"
        ],
        limits={
            "images_per_month": 400,
            "max_scale_factor": 4,
            "allowed_modes": ["block_mode", "face_mode", "waifu_mode"],
            "max_resolution": "4K",
            "batch_processing": True,
            "api_calls_per_month": 500,
            "storage_mb": 5000  # 5 GB = 5000 MB
        }
    ),
    "enterprise": SubscriptionPlan(
        id="enterprise",
        name="Enterprise",
        price=30.0,
        interval="month",
        features=[
            "800 images per month",
            "Highest upscaling quality (Block mode, Face mode, and Waifu mode)",
            "Maximum 16K output resolution",
            "Ultra-fast processing speed",
            "Batch processing",
            "API access",
            "Email support",
            "Custom integration"
        ],
        limits={
            "images_per_month": 800,
            "max_scale_factor": 16,
            "allowed_modes": ["block_mode", "face_mode", "waifu_mode"],
            "max_resolution": "16K",
            "batch_processing": True,
            "api_calls_per_month": -1,  # -1 indicates unlimited
            "storage_mb": 20000  # 20 GB = 20000 MB
        }
    )
}

class BillingHandler:
    @staticmethod
    async def get_user_billing_info(user_id: str) -> BillingInfo:
        """
        Get billing information for a user, including subscription, payment methods, and invoices.
        In a real implementation, this would fetch data from Stripe and your database.
        """
        try:
            # This is a mock implementation
            # In a real app, you would fetch this data from Stripe and your database
            
            # In a real implementation, you would check if the user exists
            # and has billing information in the database
            # For now, we'll always return mock data
            
            # Mock subscription data
            subscription = {
                "plan": "Pro",
                "status": "active",
                "renewal_date": (datetime.now().replace(day=1).strftime("%Y-%m-%d")),
                "price": "$15.00",  # Updated to match README
                "billing_cycle": "monthly",
                "features": [
                    "Unlimited images",
                    "Up to 16x upscaling",
                    "All upscaling modes",
                    "API access",
                    "Priority support",
                ]
            }
            
            # Mock payment methods
            payment_methods = [
                {
                    "id": "pm_1",
                    "type": "card",
                    "brand": "visa",
                    "last4": "4242",
                    "exp_month": 12,
                    "exp_year": 2024,
                    "is_default": True,
                }
            ]
            
            # Mock invoices
            invoices = [
                {
                    "id": "in_1",
                    "date": (datetime.now().replace(day=1).strftime("%Y-%m-%d")),
                    "amount": "$15.00",  # Updated to match README
                    "status": "paid",
                    "description": "Pro Plan - Monthly",
                    "download_url": "#",
                },
                {
                    "id": "in_2",
                    "date": (datetime.now().replace(month=datetime.now().month-1, day=1).strftime("%Y-%m-%d")),
                    "amount": "$15.00",  # Updated to match README
                    "status": "paid",
                    "description": "Pro Plan - Monthly",
                    "download_url": "#",
                },
                {
                    "id": "in_3",
                    "date": (datetime.now().replace(month=datetime.now().month-2, day=1).strftime("%Y-%m-%d")),
                    "amount": "$15.00",  # Updated to match README
                    "status": "paid",
                    "description": "Pro Plan - Monthly",
                    "download_url": "#",
                },
            ]
            
            # Mock usage data
            usage = {
                "images_processed": 87,
                "images_limit": 100,
                "api_calls": 230,
                "api_calls_limit": 500,
                "storage_used": "1.2 GB",
                "storage_limit": "5 GB",  # Matches the Pro plan's 5000 MB
            }
            
            return BillingInfo(
                subscription=subscription,
                payment_methods=payment_methods,
                invoices=invoices,
                usage=usage
            )
            
        except Exception as e:
            # Instead of raising an exception, return default billing info
            logger.error(f"Error fetching billing information: {str(e)}")
            
            # Default subscription for new users
            default_subscription = {
                "plan": "Free",
                "status": "active",
                "renewal_date": (datetime.now().replace(day=1).strftime("%Y-%m-%d")),
                "price": "$0.00",
                "billing_cycle": "monthly",
                "features": [
                    "Up to 3 images per month",
                    "2x and 4x upscaling",
                    "Basic upscaling mode",
                    "Standard support"
                ]
            }
            
            # Default usage for new users
            default_usage = {
                "images_processed": 0,
                "images_limit": 3,
                "api_calls": 0,
                "api_calls_limit": 0,
                "storage_used": "0 GB",
                "storage_limit": "0.1 GB",  # Matches the Free plan's 100 MB
            }
            
            return BillingInfo(
                subscription=default_subscription,
                payment_methods=[],
                invoices=[],
                usage=default_usage
            )
    
    @staticmethod
    async def update_subscription(user_id: str, plan_id: str) -> Dict[str, Any]:
        """
        Update a user's subscription plan.
        In a real implementation, this would update the subscription in Stripe.
        """
        try:
            # Check if the plan exists
            if plan_id not in SUBSCRIPTION_PLANS:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Invalid plan ID: {plan_id}"
                )
            
            # This is a mock implementation
            # In a real app, you would update the subscription in Stripe and your database
            
            plan = SUBSCRIPTION_PLANS[plan_id]
            
            return {
                "success": True,
                "message": f"Subscription updated to {plan.name} plan",
                "plan": {
                    "id": plan.id,
                    "name": plan.name,
                    "price": plan.price,
                    "interval": plan.interval,
                    "features": plan.features,
                    "limits": plan.limits
                }
            }
            
        except HTTPException:
            raise
        except Exception as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Error updating subscription: {str(e)}"
            )
    
    @staticmethod
    async def cancel_subscription(user_id: str) -> Dict[str, Any]:
        """
        Cancel a user's subscription.
        In a real implementation, this would cancel the subscription in Stripe.
        """
        try:
            # This is a mock implementation
            # In a real app, you would cancel the subscription in Stripe and your database
            
            return {
                "success": True,
                "message": "Subscription cancelled successfully. You will have access until the end of your billing period."
            }
            
        except Exception as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Error cancelling subscription: {str(e)}"
            )
    
    @staticmethod
    async def add_payment_method(user_id: str, payment_details: Dict[str, Any]) -> Dict[str, Any]:
        """
        Add a new payment method for a user.
        In a real implementation, this would add the payment method in Stripe.
        """
        try:
            # This is a mock implementation
            # In a real app, you would add the payment method in Stripe and your database
            
            return {
                "success": True,
                "message": "Payment method added successfully",
                "payment_method": {
                    "id": "pm_new",
                    "type": "card",
                    "brand": payment_details.get("brand", "visa"),
                    "last4": payment_details.get("last4", "4242"),
                    "exp_month": payment_details.get("exp_month", 12),
                    "exp_year": payment_details.get("exp_year", 2025),
                    "is_default": payment_details.get("is_default", False)
                }
            }
            
        except Exception as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Error adding payment method: {str(e)}"
            )
    
    @staticmethod
    async def delete_payment_method(user_id: str, payment_method_id: str) -> Dict[str, Any]:
        """
        Delete a payment method for a user.
        In a real implementation, this would delete the payment method in Stripe.
        """
        try:
            # This is a mock implementation
            # In a real app, you would delete the payment method in Stripe and your database
            
            return {
                "success": True,
                "message": "Payment method deleted successfully"
            }
            
        except Exception as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Error deleting payment method: {str(e)}"
            )
    
    @staticmethod
    async def set_default_payment_method(user_id: str, payment_method_id: str) -> Dict[str, Any]:
        """
        Set a payment method as the default for a user.
        In a real implementation, this would update the default payment method in Stripe.
        """
        try:
            # This is a mock implementation
            # In a real app, you would update the default payment method in Stripe and your database
            
            return {
                "success": True,
                "message": "Default payment method updated successfully"
            }
            
        except Exception as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Error setting default payment method: {str(e)}"
            )
    
    @staticmethod
    async def get_available_plans() -> Dict[str, Any]:
        """
        Get all available subscription plans.
        """
        try:
            plans = {}
            for plan_id, plan in SUBSCRIPTION_PLANS.items():
                plans[plan_id] = {
                    "id": plan.id,
                    "name": plan.name,
                    "price": plan.price,
                    "interval": plan.interval,
                    "features": plan.features,
                    "limits": plan.limits
                }
            
            return {
                "plans": plans
            }
            
        except Exception as e:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Error fetching subscription plans: {str(e)}"
            )
    
    @staticmethod
    async def create_checkout_session(
        user_id: str,
        plan_id: str,
        price_id: str,
        billing_cycle: str,
        success_url: str,
        cancel_url: str
    ) -> Dict[str, Any]:
        """
        Creates a Stripe checkout session for subscription.
        
        Args:
            user_id: The user ID
            plan_id: The subscription plan ID
            price_id: The Stripe price ID
            billing_cycle: The billing cycle (monthly or yearly)
            success_url: The URL to redirect to on successful payment
            cancel_url: The URL to redirect to on cancelled payment
            
        Returns:
            Dict[str, Any]: The checkout session details
        """
        try:
            logger.info(f"Creating checkout session for user {user_id} with plan {plan_id}, price {price_id}, billing cycle {billing_cycle}")
            
            # Validate the plan ID
            if plan_id not in SUBSCRIPTION_PLANS:
                logger.error(f"Invalid plan ID: {plan_id}. Available plans: {list(SUBSCRIPTION_PLANS.keys())}")
                raise ValueError(f"Invalid plan ID: {plan_id}")
            
            # Create a checkout session
            checkout_session = stripe.checkout.Session.create(
                payment_method_types=["card"],
                line_items=[
                    {
                        "price": price_id,
                        "quantity": 1,
                    },
                ],
                mode="subscription",
                success_url=success_url,
                cancel_url=cancel_url,
                metadata={
                    "user_id": user_id,
                    "plan_id": plan_id,
                    "billing_cycle": billing_cycle
                },
            )
            
            logger.info(f"Checkout session created with ID: {checkout_session.id}")
            return {
                "session_id": checkout_session.id,
                "url": checkout_session.url,
            }
        except stripe.error.StripeError as e:
            logger.error(f"Stripe error creating checkout session: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=f"Stripe error: {str(e)}"
            )
        except Exception as e:
            logger.error(f"Error creating checkout session: {str(e)}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=f"Error creating checkout session: {str(e)}"
            ) 